### PR TITLE
experimental introspection support -- Project Phantom Drift PoC

### PR DESCRIPTION
**Pull request by raulk**
### This PR represents a proof-of-concept for Project Phantom Drift (libp2p observability)

---

_See https://github.com/libp2p/go-libp2p/issues/947 for more information._

**The code being contributed herein is in EXPERIMENTAL status.** It is relatively isolated under the `introspect` package.

Test coverage of that package is 82%.

## Changes

This PR makes `BasicHost` and `RoutedHost` implement `host.IntrospectableHost`.

It contributes default implementations of:

* the introspector (`introspect.DefaultIntrospector`), responsible for introspecting the state of the system. It currently supports:
  * gathering connections
  * gathering streams
  * gathering network traffic stats
* an WebSockets-based introspection endpoint (`ws.Endpoint`), adhering to the protobuf [introspection protocol in go-libp2p-core](https://github.com/libp2p/go-libp2p-core/blob/master/introspection/pb/introspection.proto).

## Enabling introspection

Introspection MUST be enabled expressly, by using the `libp2p.Introspection` constructor option:

```go
cfg := &ws.EndpointConfig{
	ListenAddrs: []string{"127.0.0.1:"},
}

host, err := libp2p.New(ctx,
	libp2p.BandwidthReporter(metrics.NewBandwidthCounter()),
	libp2p.Introspection(
		introspect.NewDefaultIntrospector,
		ws.EndpointWithConfig(cfg),
	),
)
```

The actual listen address of the endpoint can be obtained via:

```go
h.(host.IntrospectableHost).ListenAddrs()
```

The BasicHost's constructor and `Close` method take care of managing the lifecycle of the `Introspector` and `IntrospectionEndpoint` objects, if non-nil.

---

This PR supersedes:
* https://github.com/libp2p/go-libp2p/pull/790
* https://github.com/libp2p/go-libp2p/pull/791
* https://github.com/libp2p/go-libp2p/pull/792
* https://github.com/libp2p/go-libp2p/pull/885

## Taking it for a spin

Check out the REPL: https://github.com/libp2p/repl. These changes have been integrated in the REPL to enable demos.

## TODO

- [ ] Make `IntrospectionEndpoint` restartable.
- [ ] Make start and close idempotent.

### CODE REVIEW

File: p2p/host/basic/basic_host.go [hunk 1]
```diff
@@ -204,6 +220,7 @@ func NewHost(ctx context.Context, n network.Network, opts *HostOpts) (*BasicHost
 		h.mux = opts.MultistreamMuxer
 	}
 
+	// Start ID Service
```
Comment (raulk) [inline p2p/host/basic/basic_host.go:223]: bad line

File: introspect/event_manager.go [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+)
+
+var (
+	jsType     = reflect.TypeOf(new(event.RawJSON)).Elem()
+	peerIdType = reflect.TypeOf(new(peer.ID)).Elem()
+	timeType   = reflect.TypeOf(new(time.Time)).Elem()
+	maddrType  = reflect.TypeOf(new(multiaddr.Multiaddr)).Elem()
+)
+
+type eventManager struct {
```
Comment (raulk) [inline introspect/event_manager.go:27]: needs godocs

File: introspect/default_inspector.go [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+		reporter:     reporter,
+		started:      time.Now(),
+	}
+
+	return d, nil
+}
+
+func (d *DefaultIntrospector) Close() error {
+	if err := d.wsub.Close(); err != nil {
+		return fmt.Errorf("failed while trying to close wildcard eventbus subscription: %w", err)
```
Comment (vyzo) [inline introspect/default_inspector.go:59]: shouldn't we still close the rest of resources (and maybe collect a multierror)?

File: introspect/default_inspector.go [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+
+	return nil
+}
+
+func (d *DefaultIntrospector) FetchRuntime() (*pb.Runtime, error) {
+	return &pb.Runtime{
+		Implementation: "go-libp2p",
+		Platform:       runtime.GOOS,
+		PeerId:         d.host.ID().Pretty(),
+		Version:        "",
```
Comment (vyzo) [inline introspect/default_inspector.go:73]: why is the version empty?

File: introspect/event_manager.go [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+	defer em.Unlock()
+
+	if em.closed {
+		em.closeWg.Wait()
+		return nil
+	}
+
+	close(em.closeCh)
+	em.closeWg.Wait()
+	return nil
```
Comment (vyzo) [inline introspect/event_manager.go:65]: shouldn't we also set `em.closed` to true?

File: introspect/default_inspector.go (hunk 3 above)
Comment (raulk) [reply introspect/default_inspector.go:59]: Yup, let's do that..

File: introspect/default_inspector.go (hunk 4 above)
Comment (raulk) [reply introspect/default_inspector.go:73]: We don't really have a go-libp2p version/commit anyway to send back, as far as I'm aware?

File: introspect/event_manager.go (hunk 5 above)
Comment (raulk) [reply introspect/event_manager.go:65]: Yup. I have a TODO item to tackle closure idempotency. Will take care of that in a follow-up commit before merging.

**Comment by raulk** [comment]
@vyzo yup, this code is certainly not bug-free! It's marked as experimental, and these components are only used if the user enables the feature, so I think we're safe.

File: introspect/default_inspector.go (hunk 4 above)
Comment (vyzo) [reply introspect/default_inspector.go:76]: maybe use the user agent?

**Comment by marten-seemann** [comment]
Closing this PR. See https://github.com/libp2p/go-libp2p/issues/775#issuecomment-1405923140 for motivation.

**Pull request closed without merging**